### PR TITLE
[New features] Add hash routing and softplus score functions

### PR DESCRIPTION
**Pull request by adam-xiaoyao**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

1. **Hash Routing**：在 `TopKRouter` 中新增基于 `tid2eid` 查表的 hash routing 分支。当 `config.moe_n_hash_layers > 0` 时，层号 `< moe_n_hash_layers` 的前几层 MoE layer 使用 hash 路由，评分权重仍由 gate logits 计算。`TransformerConfig` 新增 `moe_n_hash_layers`（int，默认 0）和 `actual_vocab_size`（int|None）字段，并在 `__post_init__` 中加了合法性校验。
2. **Softplus 评分函数**：新增 `sftplus`（`F.softplus`）和 `sqrtsoftplus`（`sqrt(softplus)`）两种 `scoring_func` 选项。
3. **MTP 层隔离**：MTP layer 通过 `is_mtp_layer=True` 禁用 hash routing，避免 MTP 层误触 hash 路由逻辑。
4. 删除了 `TransformerConfig` 中的 3 个重复字段定义：`apply_residual_connection_post_layernorm`、`activation_func_clamp_value`、`glu_linear_offset`（这些字段在文件其他位置已有定义）。

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/transformer_config.py [hunk 1]
```diff
@@ -229,17 +229,6 @@ class TransformerConfig(ModelParallelConfig):
     """Offset term in the GLU activation function: activation_func(x[0]) * (x[1] + offset). Only
     used when gated_linear_unit is True"""
```
Comment (adam-xiaoyao) [inline src/paddlefleet/transformer/transformer_config.py:231]: 和221 ~ 228的重复。

File: src/paddlefleet/transformer/transformer_config.py [hunk 2]
```diff
@@ -462,6 +453,16 @@ class TransformerConfig(ModelParallelConfig):
     moe_router_force_load_balancing: bool = False
     """Force load balancing with random logits for MoE router."""
 
+    moe_n_hash_layers: int = 0
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:456]: 这个不需要， arguments.py和yaml_arguments.py是循环展开的。

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 3]
```diff
@@ -142,6 +142,7 @@ def __init__(
         config: TransformerConfig,
         sublayers: MoESublayers | None = None,
         pg_collection: ProcessGroupCollection | None = None,
+        layer_number: int | None = None,
     ):
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_layer.py:145]: OK, 我修改。

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 4]
```diff
@@ -1336,4 +1337,6 @@ def set_layer_number(self, layer_number):
         assert hasattr(self.gate, "set_layer_number"), (
             "expect gate has method 'set_layer_number'"
         )
+        # Hash routing activation (moe_n_hash_layers) is decided by the router
+        # itself based on layer_number. See TopKRouter._setup_hash_layer.
         self.gate.set_layer_number(layer_number)
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_layer.py]: OK, 我修改下。

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 5]
```diff
@@ -758,14 +830,84 @@ def _call_topk_method(
     def set_layer_number(self, layer_number):
         self.layer_number = layer_number
 
+    def _setup_hash_layer(self, layer_number, is_mtp_layer: bool = False):
+        """Activate hash routing for this layer if it falls in the hash range.
+
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py:837]: OK， 我修改下。

File: src/paddlefleet/transformer/transformer_config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:456]: 不需要。 arguments.py和yaml_arguments.py是循环展开的。

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 6]
```diff
@@ -732,6 +741,69 @@ def _topk_noaux_tc(
...
+        """
+        if self.tid2eid is None:
+            raise ValueError(
+                "tid2eid buffer is not registered; hash routing is not initialized."
+            )
+        score_function = self.scoring_func
+        orig_dtype = logits.dtype
+        logits_fp32 = logits.cast("float32")
+        if score_function == "softmax":
+            scores = F.softmax(logits_fp32, axis=-1).cast(orig_dtype)
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py:777]: OK, 我按建议来。

File: src/paddlefleet/transformer/transformer_config.py [hunk 7]
```diff
@@ -462,6 +453,16 @@ class TransformerConfig(ModelParallelConfig):
...
+    """Number of leading transformer layers that use hash-based MoE routing.
+    Layers with layer_number < moe_n_hash_layers (0-indexed) use a pre-computed
+    tid2eid lookup table for expert selection instead of learned top-k routing.
+    Score weights are still computed from the gate logits. 0 disables hash routing."""
+
+    actual_vocab_size: int | None = None
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:462]: 不需要

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 8]
```diff
@@ -807,6 +959,43 @@ def forward(self, input, input_ids=None):
...
+            mask = (probs > 0).cast(logits.dtype)
+
+            # Apply padding (input_ids == 0):
+            # routing_map = routing_map & ~padding_mask.
+            # input_ids_none_zero_mask shape: [N, 1], broadcast over expert/topk dim.
+            if input_ids_none_zero_mask is not None:
+                valid_mask = input_ids_none_zero_mask.cast(mask.dtype)
+                mask = mask * valid_mask
+                probs = probs * valid_mask
+                top_gate = top_gate * valid_mask
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py:992]: 已经修复

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 9]
```diff
@@ -755,17 +829,87 @@ def _call_topk_method(
             raise NotImplementedError(f"Invalid topk_method: {topk_method}")
         return top_gate, top_idx
 
-    def set_layer_number(self, layer_number):
+    def set_layer_number(self, layer_number, is_mtp_layer: bool = False):
         self.layer_number = layer_number
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_router.py:833]: 没必要添加， 暂时先这样。

File: src/paddlefleet/transformer/transformer_config.py [hunk 10]
```diff
@@ -364,7 +353,9 @@ class TransformerConfig(ModelParallelConfig):
     """Number of experts to route to for each token."""
 
     scoring_func: str = "softmax"
-    """Score function for MoE routing. Can be "softmax" or "sigmoid"."""
+    """Score function for MoE routing. Options: "softmax", "sigmoid", "tanh",
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/transformer_config.py:356]: 文档已较为详细， 暂不添加。

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by From00 — APPROVED** [review]
LGTM
